core/templates/services/playthrough.service.ts

```diff
@@ -171,16 +176,19 @@ class MultipleIncorrectAnswersTracker {
   providedIn: 'root'
 })
 export class PlaythroughService {
-  private explorationId: string = null;
-  private explorationVersion: number = null;
-  private learnerIsInSamplePopulation: boolean = null;
-
-  private eqTracker: EarlyQuitTracker = null;
-  private cstTracker: CyclicStateTransitionsTracker = null;
-  private misTracker: MultipleIncorrectAnswersTracker = null;
-  private recordedLearnerActions: LearnerAction[] = null;
-  private playthroughStopwatch: Stopwatch = null;
-  private playthroughDurationInSecs: number = null;
+  // These properties are initialized using Angular lifecycle hooks
+  // and we need to do non-null assertion, for more information see
+  // https://github.com/oppia/oppia/wiki/Guide-on-defining-types#ts-7-1
+  private explorationId!: string;
+  private explorationVersion!: number;
+  private learnerIsInSamplePopulation!: boolean;
+
+  private eqTracker: EarlyQuitTracker | null = null;
+  private cstTracker: CyclicStateTransitionsTracker | null = null;
+  private misTracker: MultipleIncorrectAnswersTracker | null = null;
+  private recordedLearnerActions: LearnerAction[] | null = null;
+  private playthroughStopwatch: Stopwatch | null = null;
+  private playthroughDurationInSecs: number | null = null;
```

Comment:
should we have comments here on why each of these properties are null?

Reply:
Done